Document frontend translation syntax.

umairwaheed opened the pull request:
Fixes: #1090

@timabbott please review.

smarx commented:
*Automated message from Dropbox CLA bot*

@umairwaheed, it looks like you've already signed the Dropbox CLA. Thanks!

timabbott commented on docs/translating.md:
```diff
@@ -71,7 +71,8 @@ To ensure we always internationalize our JSON errors messages, the
 
 ## Frontend Translations
-The first step in translating the frontend is to create the translation
+Zulip uses [i18next][] library for the frontend translation. To make the
```
s/uses/uses the/

timabbott commented on docs/translating.md:
```diff
@@ -71,7 +71,8 @@ To ensure we always internationalize our JSON errors messages, the
 ## Frontend Translations
-The first step in translating the frontend is to create the translation
+Zulip uses [i18next][] library for the frontend translation. To make the
+frontend translation work you first need to create the translation
```
s/translation/translations`

timabbott commented on docs/translating.md:
```diff
@@ -84,12 +85,85 @@ The second step is to upload the translatable strings to Transifex using
+            "keyWithCount_plural": "__count__ items"
+        }
+    }
+}
```
Does making this work mean we'll need to hand-edit the translations.json file for English?  Seems like that might be an annoying workflow given that `manage.py makemessages` will overwrite it...

umairwaheed replied:
Yeah, that's a problem, didn't think of that. How about we generate plurals using `manage.py makemessages` as well?

umairwaheed replied:
I think there are two ways in which we can do this, one is to create a new syntax to make the parser know which translation strings should have plurals, second is to guess from the content of the strings.

umairwaheed commented:
@timabbott, done.

brainwane commented:
It would be great to properly capitalize proper nouns throughout, such as:

Unicode
JavaScript

brainwane commented on docs/translating.md:
```diff
@@ -84,12 +85,85 @@ The second step is to upload the translatable strings to Transifex using
+When you run `python manage.py makemessages`, Zulip parses all the code
+for strings which are marked for translation and dumps these strings in
+a translation resource which is uploaded to the Transifex. Zulip provides
+syntax for marking strings in both Javascript files and Handlebar
```
What is a "translation resource"? Could you link to or include a quick explanation/definition?

umairwaheed replied:
I have added the explanation.

brainwane commented on docs/translating.md:
```diff
@@ -71,7 +71,8 @@ To ensure we always internationalize our JSON errors messages, the
+the frontend translations work you first need to create the translation
 files using `python manage makemessages`. This command will create
 translation files under `static/locale`, the location can be changed by
 passing an argument to the command, however make sure that the location is
```
This should be separated into three sentences.

umairwaheed replied:
Done.

brainwane commented on docs/translating.md:
```diff
@@ -84,12 +85,85 @@ The second step is to upload the translatable strings to Transifex using
 The final step is to get the translated files from Transifex using
 `tx pull -a`.
 
```
What's the end result? What are some example filenames & file extensions I'll see?

umairwaheed replied:
I have added some explanation.

brainwane commented on docs/translating.md:
````diff
@@ -84,12 +85,85 @@ The second step is to upload the translatable strings to Transifex using
+```
+
+Note the syntax for declaring a variable, we enclose the variables with
+`__` instead of `{{ ... }}` because we need to differentiate the variables
````
Instead of a comma here, use a colon, like:

> Note the syntax for declaring a variable:

umairwaheed replied:
Done

umairwaheed commented:
@brainwane, updated. Please review again. Thanks.

brainwane commented on docs/translating.md:
```diff
@@ -76,25 +76,104 @@ To ensure we always internationalize our JSON errors messages, the
+### Marking Strings for Translation
+When you run the `python manage.py makemessages` command, Django parses
+the Zulip code to find the translatable strings and dumps these strings
+in a file. This file is called the resource file. The resource file is
```
Usually, what's the file extension of a resource file?

timabbott replied:
`.po`

brainwane replied:
I've just looked in `static/locale/` and looked at the individual language directories. If we have one resource file per language and the resource file is always `translations.json` it would be great to say that in this document.

umairwaheed replied:
Sure.

brainwane commented on docs/translating.md:
```diff
@@ -76,25 +76,104 @@ To ensure we always internationalize our JSON errors messages, the
 
-The final step is to get the translated files from Transifex using
-`tx pull -a`.
+The rules for plurals are same as for the Javascript files. You just have
```
Please use "JavaScript" here.

brainwane commented on docs/translating.md:
```diff
@@ -76,25 +76,104 @@ To ensure we always internationalize our JSON errors messages, the
+the `tx pull -a` command. This command will download the resource files
+from Transifex and replace your local resource files with them.
+
+### Marking Strings for Translation
```
I appreciate the step-by-step instructions. 

Since these files should be tracked in git, perhaps we should instruct people in that. If I'm a Zulip developer, and I've made user-visible changes in Zulip, I should make sure to run `python manage.py makemessages` and `git add` those files before I make a pull request, right?

Am I right that `python manage makemessages` will add new messages to the existing `translations.json` files, and so it's safe to re-run it multiple times as I work?

umairwaheed replied:
>Am I right that python manage makemessages will add new messages to the existing translations.json files, and so it's safe to re-run it multiple times as I work?

Yep, the command is idempotent.

umairwaheed replied:
>Since these files should be tracked in git, perhaps we should instruct people in that. If I'm a Zulip developer, and I've made user-visible changes in Zulip, I should make sure to run python manage.py makemessages and git add those files before I make a pull request, right?

I think the person who is responsible for uploading and downloading the resources from the Transifex should update the Git repository once new translations are available. Committing the `translations.json` file in a new PR won't serve any immediate purpose.

brainwane replied:
> I think the person who is responsible for uploading and downloading the resources from the Transifex should update the Git repository once new translations are available. Committing the translations.json file in a new PR won't serve any immediate purpose.

@timabbott currently, is there an automated process for noticing when new translations are available from Transifex?

brainwane commented on docs/translating.md:
```diff
@@ -76,25 +76,104 @@ To ensure we always internationalize our JSON errors messages, the
+the Zulip code to find the translatable strings and dumps these strings
+in a file. This file is called the resource file. The resource file is
+then uploaded to the Transifex using `tx push -s -a` command. To mark a
+string translatable in the Javascript files pass it to the `i18n.t` function.
```
This is a small thing, but I'd prefer it if you used "JavaScript" here.

umairwaheed replied:
Sure :)

brainwane commented on docs/translating.md:
```diff
@@ -76,25 +76,104 @@ To ensure we always internationalize our JSON errors messages, the
+Note: In the second example above, instead of enclosing the variable with
+handlebars `{{ }}`, we enclose the it with `__` because we need to
+differentiate the variable from the Handlebar tags. The symbol which is
+used to enclose the variables can be changed in `/static/js/src/main.js`.
```
Thank you for this clear explanation (the differentiation)!

brainwane commented on docs/translating.md:
```diff
@@ -76,25 +76,104 @@ To ensure we always internationalize our JSON errors messages, the
+
+For further reading on plurals, read the [official] documentation.
+
+To mark the strings as translatble in Handlebar templates, Zulip registers
```
"translatable" is misspelled here. :)

brainwane commented:
@umairwaheed Thanks for the update! Now that I better understand what a translation resource is, I've been able to provide some more substantive and useful questions and feedback. Looking forward to your next update!

umairwaheed commented:
@brainwane, thanks for the review! I have updated the documentation according to the feedback. @timabbott, I have also updated the `makemessages` command so that it is now aware of the plurals; it will not remove the plural keys on re-runs.

brainwane commented on docs/translating.md:
```diff
@@ -76,25 +76,111 @@ To ensure we always internationalize our JSON errors messages, the
+Zulip code anymore; it will only delete plural keys, see below for the
+documentation on plural translations, when the corresponding singular
+key is absent, and it will not override the value of a singular key if
+that value contains a translation.
```
I think you're saying that `makemessages` deletes singular keys if they aren't used in source code anymore, right?

Here's a suggested revision, to fix the punctuation and grammar:

Note: the `makemessages` command is idempotent in that:
* it will only delete singular keys in the resource file when they are no longer used in Zulip code
* it will only delete plural keys (see below for the documentation on plural translations) when the corresponding singular key is absent
* it will not override the value of a singular key if that value contains a translation

umairwaheed replied:
Cool.

brainwane commented on docs/translating.md:
```diff
@@ -76,25 +76,111 @@ To ensure we always internationalize our JSON errors messages, the
+Zulip uses the [i18next][] library for the frontend translations. To make
+the frontend translations work, the first step is to create the JSON
+formatted [resource][] files using the `python manage makemessages`
+command. This command will create a resource file called `translations.json`
```
I'm actually a little confused now, sorry. What is the relationship between, for instance, `locale/es/LC_MESSAGES/django.po` and `static/locale/es/translations.json`? Which one should we be calling "the resource file", and what is a term for the other file?

timabbott replied:
The former is for the backend and the latter for the frontend.  They're both resource files.  

@umairwaheed what do you think about moving them all to live under `static/locale`?  Seems like that might be a less confusing story :)

umairwaheed replied:
> what do you think about moving them all to live under static/locale? Seems like that might be a less confusing story :)

Yeah, I think that would be better. I'll try to figure this out.

umairwaheed replied:
>I'm actually a little confused now, sorry. What is the relationship between, for instance, locale/es/LC_MESSAGES/django.po and static/locale/es/translations.json? Which one should we be calling "the resource file", and what is a term for the other file?

@brainwane, I'll highlight this difference in the docs.

umairwaheed replied:
@timabbott, there is one caveat, moving the `.po` files to `static/locale` will make them publicly accessible.

umairwaheed commented:
Sorry for the delay on this. I have made some modifications. Please review. Thanks.

@timabbott, I have also moved the backend translation resources to `static/locale`.

umairwaheed commented:
The build seems to be failing. I will investigate later.

brainwane commented on docs/translating.md:
```diff
@@ -19,21 +19,69 @@ project or add a language so that we can be sure to accept your
+The config file that maps the resources from Zulip to Transifex is
+located at `.tx/config`. Django  recognizes `zh_CN` instead of `zh-HANS`
+for simplified Chinese language (this is fixed in Django 1.9). This
+idiosyncrasy is also handled in the Transifex config file.
```
Nice usage of "idiosyncrasy"!

brainwane commented on docs/translating.md:
```diff
@@ -19,21 +19,69 @@ project or add a language so that we can be sure to accept your
+
+## Translation Process
+
+The end-to-end process to the get the translations working is as
```
Extraneous "the" here. Should be:

> The end-to-end process to the get the translations working

brainwane commented:
@umairwaheed you've clarified things a lot. Aside from the extraneous "the" I mention in my 1 comment above, I think this is good to go from a documentation perspective. Thanks.

umairwaheed commented:
@brainwane thanks. I have removed the extraneous 'the'. @timabbott, please review the changes related to the code. Thanks.

umairwaheed commented:
@timabbott, please don't merge the last commit, I just added to make that i18n test pass.

timabbott commented:
Merged (minus that last commit), thanks @umairwaheed!

timabbott commented:
@umairwaheed the zh_CN/zh-CN thing now looks really funny with these under the same tree (and transifex now pulls down the files for either for both, which doesn't seem right).  I had to merge 3c591aa72436a644317f4ab1e8e7676a87772245 to fix an issue with language_options.json having duplicates; but I think more is probably needed here to make the new path nice.

umairwaheed commented:
Cool,  thanks. I'll look into this as well.
